presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,775 @@
+import static com.facebook.presto.execution.buffer.BufferState.NO_MORE_BUFFERS;
+import static com.facebook.presto.execution.buffer.BufferState.OPEN;
+import static com.facebook.presto.execution.buffer.OutputBuffers.BufferType.SPOOLING;
+import static com.facebook.presto.spi.StandardErrorCode.SPOOLING_STORAGE_ERROR;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readSerializedPages;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterators.advance;
+import static com.google.common.collect.Range.closedOpen;
+import static com.google.common.util.concurrent.Futures.catchingAsync;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.Futures.transform;
+import static com.google.common.util.concurrent.Futures.transformAsync;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+@SuppressWarnings("UnstableApiUsage")
+public class SpoolingOutputBuffer
+        implements OutputBuffer
+{
+    private final TaskId taskId;
+    private final String taskInstanceId;
+    private final OutputBuffers outputBuffers;
+    private final StateMachine<BufferState> state;
+    private final TempDataOperationContext tempDataOperationContext;
+    private final TempStorage tempStorage;
+    private final DataSize threshold;
+    private final FinalizerService finalizerService;
+    private final ListeningExecutorService executor;
+
+    private final AtomicLong totalBufferedBytes = new AtomicLong();
+    private final AtomicLong totalBufferedPages = new AtomicLong();
+    private final AtomicLong totalPagesAdded = new AtomicLong();
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+
+    private final OutputBufferId outputBufferId = new OutputBufferId(0);
+
+    private static final Logger log = Logger.get(SpoolingOutputBuffer.class);
+
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+    private final AtomicLong currentMemorySequenceId = new AtomicLong();
+    private final AtomicLong currentSequenceId = new AtomicLong();
+    private final AtomicLong startPage = new AtomicLong();
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+    private final AtomicLong totalInMemoryBytes = new AtomicLong();
+    private final AtomicLong peakMemoryUsage = new AtomicLong();
+
+    private final AtomicLong totalStorageBytesAdded = new AtomicLong();
+    private final AtomicLong totalStoragePagesAdded = new AtomicLong();
+
+    @GuardedBy("this")
+    private final Deque<HandleInfo> handleInfoQueue = new LinkedList<>();
```

Comment:
We may need to limit this queue to avoid potential memory leak. Do we anything that will allow us to the back pressure when the temp file system is slow? cc: @tdcmeehan  ?